bootstrap.rkt

```diff
@@ -0,0 +1,476 @@
+;            (let [quadruple (generate-quasiquote (,a ,b ,c ,d) stx)]
+;              (pure (generate-quasiquote '(four-elements ,quadruple) stx)))]
+;           [(keyword ,tail ...)
+;            (pure (generate-quasiquote '(keyword-prefixed (,tail ...)) stx))]))])))
+; =>
+; `(define-macros
+;    ([my-macro
+;      (lambda (stx)
+;        ,(generate-syntax-case 'my-macro 'stx (list 'keyword)
+;          (list
+;            (list '()
+;                  `(pure ''nil))
+;            (list '((,a ,b) (,c ,d))
+;                  `(let [quadruple ,(generate-quasiquote '(,a ,b ,c ,d) 'stx)]
+;                     (pure ,(generate-quasiquote ''(four-elements ,quadruple) 'stx))))
+;            (list '(keyword ,tail ...)
+;                  `(pure ,(generate-quasiquote ''(keyword-prefixed (,tail ...)) 'stx))))))]))
+(define-syntax (auto-splice stx)
+  (racket-syntax-case stx (generate-quasiquote generate-syntax-case)
+    [(_ ())
+     #''()]
+    [(_ (generate-quasiquote pat stx-name))
+     #'(generate-quasiquote 'pat 'stx-name)]
+    [(_ (generate-syntax-case macro-name stx-expr (keyword ...)
+          [lhs rhs ...] ...))
+     #'(generate-syntax-case 'macro-name 'stx-expr (list 'keyword ...)
+         (list
+           (list 'lhs
+                 (auto-splice rhs) ...)
+           ...))]
+    [(_ (head tail ...))
+     #'(cons (auto-splice head)
+             (auto-splice (tail ...)))]
+    [(_ x)
+     #''x]))
+
+(void
+  (call-with-output-file
+    "examples/dot-dot-dot.kl"
+    #:exists 'truncate
+    (lambda (out)
+      (parameterize ([current-output-port out]
+                     [pretty-print-columns 40])
+        (let ([newline
+                (lambda ()
+                  (pretty-print-newline out (pretty-print-columns)))]
+              [displayln
+                (lambda (s)
+                  (pretty-display s #:newline? #t))]
+              [writeln
+                (lambda (v)
+                  (pretty-write v #:newline? #t))])
+          (displayln "#lang \"prelude.kl\"")
+          (displayln "-- GENERATED BY ../bootstrap.rkt, DO NOT EDIT")
+
+          (map
+            (lambda (form)
+              (newline)
+              (writeln form))
+            (list
+              '(import (rename "prelude.kl"
+                               [syntax-case raw-syntax-case]))
+              '(import (rename (shift "prelude.kl" 1)
+                               [syntax-case raw-syntax-case]))
+              '(import (shift "identifier.kl" 1))
+              '(import (shift "list-syntax.kl" 1))
+              '(import (shift "temporaries.kl" 1))
+
+              (auto-splice
+                (define-macros
+                  ([fancy-unquote (lambda (stx) (syntax-error '"unquote used out of context" stx))]
+                   [fancy-...     (lambda (stx) (syntax-error '"... used out of context" stx))]
+                   [fancy-_       (lambda (stx) (syntax-error '"_ used out of context" stx))]
+                   [fancy-syntax-case
```

Comment:
This definition of `fancy-syntax-case` is almost the same as the definition of `generate-syntax-case`, except is using the Klister names for things instead of the Racket names. I considered writing a macro which would allow me to write the code once in some language-agnostic way and then specialize it to both Racket and Klister. I'm still not sure if that would make the code simpler or more complicated.